# refLinker
Chromosome-length haplotype determination from Hi-C and external reference panels

## Installation

`refLinker` is available as a compiled release: https://github.com/gbrunette/refLinker/releases/tag/v1.0 and as a Docker image: https://hub.docker.com/r/gregbrun/reflinker_compile_amd64

Alternatively, we provide the steps for building `refLinker` below:

## Building `refLinker`

This code requires bamtools, htslib, c++11, and zlib libraries.

htslib: https://github.com/samtools/htslib
bamtools: https://github.com/pezmaster31/bamtools


From the linker directory run build.sh or install manually:

Installing htslib locally

```
git clone https://github.com/samtools/htslib
cd htslib
autoheader
autoconf
./configure --prefix=/path/to/linker/packages/htslib/
make; make install
cd ..
```

Installing bamtools locally

```
git clone git://github.com/pezmaster31/bamtools.git
cd bamtools
mkdir build; cd build
cmake -DCMAKE_INSTALL_PREFIX=/path/to/linker/packages/bamtools/ ..
make; make install
cd ../..
```

Make an empty output directory if not present:

```
mkdir ./output
```

Building `refLinker`:

```
make
```

# Description

`refLinker` resolves complete chromosomal haplotypes using Hi-C data and statistical phasing methods. 

`refLinker` performs haplotype refinement on an initial haplotype "guess," here, the statistical phasing haplotype generated by `EAGLE2` (https://alkesgroup.broadinstitute.org/Eagle/). 

`refLinker` then introduces perturbations to linkage assignment that maximize the number of observed intra-homolog Hi-C contacts using steepest descent approaches. 

![image](https://github.com/user-attachments/assets/87fb40a0-7f0d-407a-a767-5e7d2035dfe6)



The resulting linkage assignment represents the whole-chromosome haplotype. Steps to run `refLinker` are outlined below.

## Extracting Hi-C links

Starting from a statistical phasing haplotype, `{statistical haplotype}.vcf.gz` and aligned Hi-C reads, `{Hi-C reads}.bam`, extracts Hi-C links between variant genotypes for a single chromosome with the following command:

`./linker extract -v {statistical haplotype}.vcf.gz -i {Hi-C reads}.bam -n {sample name} -e hic -c {chrom}`

This generates map from reads to variants, written to the file `graph_variant_{sample name}_{chrom}.dat`

## Haplotype refinement

Chromosome-length haplotype inference is then performed using the reads-to-variants map and statistical phasing haplotype with the following command:

`./linker pop -v {statistical haplotype}.vcf.gz -g graph_variant_{sample name}_{chrom}.dat -c {chrom} -e -10.0 -w 2000 -p 0.999`

Where the arguments `-e` and `-p` specify the block-flipping penalty and linkage pruning cut-offs respectively. `reflinker pop` writes whole-chromosome haplotype to the file `pop_hap_solution_{sample name}_{chrom}.dat`, where column 1 corresponds to variant position, column 4 corresponds to refLinker haplotype, and column 5 corresponds to the initial (e.g. EAGLE2) haplotype assignment.  

## Haplotype recovery

The python script `eagle2_recover.py` calculates the haplotype linkage between common, unlinked variants omitted from the refLinker Hi-C scaffold, based on their EAGLE2 linkage to the closest linked variant within 5 kb.
